share/eds/store.go

```diff
@@ -262,17 +264,35 @@ func (s *Store) put(ctx context.Context, root share.DataHash, square *rsmt2d.Ext
 		return fmt.Errorf("failed to initiate shard registration: %w", err)
 	}
 
+	var result dagstore.ShardResult
 	select {
+	case result = <-ch:
 	case <-ctx.Done():
-		// if context finished before result was received, track result in separate goroutine
+		// if the context finished before the result was received, track the result in a separate goroutine
 		go trackLateResult("put", ch, s.metrics, time.Minute*5)
 		return ctx.Err()
-	case result := <-ch:
-		if result.Error != nil {
-			return fmt.Errorf("failed to register shard: %w", result.Error)
-		}
-		return nil
 	}
+
+	if result.Error != nil {
+		return fmt.Errorf("failed to register shard: %w", result.Error)
+	}
+
+	// the accessor returned in the result will be nil, so the shard needs to be acquired first to
+	// become available in the cache. It might take some time, and the result should not affect the put
+	// operation, so do it in a goroutine
```

Comment:
>So let's look into what does make this some time. We already indexed everything at this point and we have in memory mount, so acquiring should not take a lot or do I miss anything?

It is happening, because dagstore uses single event loop, that causes everything to get stuck, once some operation obtained the control and not releasing it for some time. 


The motivation for this change was:
This operation sometimes happens to be long sometimes based on observed data for long hanging operations. Usually it is fast, but could take 10-50s to finish in times when storage is doing a lot of Put operations (while historical sampling). The extra delay to acquire the shard is putting Put operation into risk of context timeout error. If error happens, it will cause retry of whole sampling + storing which is heavy. So to remove this risk, accessor for the cache is acquired in parallel, allowing Put to be released right after shard is registered successfully. 

![image](https://github.com/celestiaorg/celestia-node/assets/13818348/d957a11a-1f8b-4645-b9d1-279bca113340)


On your downsides, I think they are neglect able compared to errors that would be caused by sync cache registration:
 -  Tests with no sleeps.
 -> Only tests related to cache needs a sleep. There are not that many
 - Guarantee that there are no errors in acquiring.
 -> I don't see how async cache registration bring any risk of acquiring error. Would be great if you can add examples of such errors introduced by async.
 - Less implementation complexity
 -> It not that complex, with just extra "fire and forget" go-routine semantic being 4 lines of code
            
>Orthogonal thought: Why aren't we calling RegisterShard in parallel with WriteEDS?

Because if register shard fails, we won't be able to acquire it later. So the registration have to be done sync to allow transactional property of writeeds+registerShard

Reply:
Is there a way to construct a home-backed accessor using the data we have in put to avoid interactions with DagStore event loop? Theoretically seems possible

Reply:
Although, we may just leave with this as we will stop using dagstore in a few months from now

Reply:
It is possible, but it will have to be different accessor type, unrelated to dagstore. It would require to make and adapter for inMemoryMount to satisfy Accessor interface. Such accessor will not be tracked by dagstore status model and refs count tho. It is a bit tricky, so definitely not in the scope of this PR.
My take is to keep it as is, and spent more time working on dagstore removal

Reply:
👍🏻